nmap.go

```diff
@@ -70,353 +70,143 @@ func NewScanner(options ...Option) (*Scanner, error) {
-		done <- cmd.Wait()
+		err := cmd.Wait()
+		if streamerGroup != nil {
+			streamerError := streamerGroup.Wait()
+			if streamerError != nil {
+				*warnings = append(*warnings, errors.WithMessage(err, "read from stdout failed").Error())
+			}
+		}
+		done <- err
 	}()
 
 	// Make goroutine to check the progress every second.
 	// Listening for channel doneProgress.
-	go func() {
-		type progress struct {
-			TaskProgress []TaskProgress `xml:"taskprogress" json:"task_progress"`
-		}
-		p := &progress{}
-		for {
-			select {
-			case <-doneProgress:
-				close(liveProgress)
-				return
-			default:
-				time.Sleep(time.Second)
-				_ = xml.Unmarshal(stdout.Bytes(), p)
-				//result, _ := Parse(stdout.Bytes())
-				if len(p.TaskProgress) > 0 {
-					liveProgress <- p.TaskProgress[len(p.TaskProgress)-1].Percent
-				}
+	if s.liveProgress != nil {
+		go func() {
+			type progress struct {
+				TaskProgress []TaskProgress `xml:"taskprogress" json:"task_progress"`
 			}
-		}
-	}()
-
-	// Wait for nmap process or timeout.
-	select {
-	case <-s.ctx.Done():
-		// Trigger progress function exit.
-		close(doneProgress)
-
-		// Context was done before the scan was finished.
-		// The process is killed and a timeout error is returned.
-		_ = cmd.Process.Kill()
-
-		return nil, warnings, ErrScanTimeout
-	case <-done:
-		// Trigger progress function exit.
-		close(doneProgress)
-
-		// Process nmap stderr output containing none-critical errors and warnings.
-		// Everyone needs to check whether one or some of these warnings is a hard issue in their use case.
-		if stderr.Len() > 0 {
-			warnings = strings.Split(strings.Trim(stderr.String(), "\n"), "\n")
-		}
-
-		// Check for warnings that will inevitably lead to parsing errors, hence, have priority.
-		if err := analyzeWarnings(warnings); err != nil {
-			return nil, warnings, err
-		}
-
-		// Parse nmap xml output. Usually nmap always returns valid XML, even if there is a scan error.
-		// Potentially available warnings are returned too, but probably not the reason for a broken XML.
-		result, err := Parse(stdout.Bytes())
-		if err != nil {
-			warnings = append(warnings, err.Error()) // Append parsing error to warnings for those who are interested.
-			return nil, warnings, ErrParseOutput
-		}
-
-		// Critical scan errors are reflected in the XML.
-		if result != nil && len(result.Stats.Finished.ErrorMsg) > 0 {
-			switch {
-			case strings.Contains(result.Stats.Finished.ErrorMsg, "Error resolving name"):
-				return result, warnings, ErrResolveName
-			// TODO: Add cases for other known errors we might want to guard.
-			default:
-				return result, warnings, fmt.Errorf(result.Stats.Finished.ErrorMsg)
+			p := &progress{}
+			for {
+				select {
+				case <-doneProgress:
+					close(s.liveProgress)
+					return
+				default:
+					time.Sleep(time.Second)
+					_ = xml.Unmarshal(stdout.Bytes(), p)
+					if len(p.TaskProgress) > 0 {
+						s.liveProgress <- p.TaskProgress[len(p.TaskProgress)-1].Percent
```

Comment:
It would be cleaner to create a var such as `progressIdx := len(p.TaskProgress) - 1` and use the var later to increase readability IMO, WDYT?